Linearise more shaders, updated Lottes, reworked volumetric fog

TheFireKahuna opened the pull request:
- Updated VTLottes with vtastek's new changes
- Optimised VTLottes by precalcing shader values DLL-side 
- Linearised Exterior/Interior Shadows, Ambient Occlusion, Underwater, Water, Rain, Snow Accumulation and Wet World
- Partially linearised Image Adjust
- Linearised bloom blending, exponents should ideally be sub 1 to avoid artefacts 
- Fixed Godrays, had too much linearisation
- Moved AO to after Shadows as linear space blending is different (needs testing)
- Reworked Volumetric Fog with the unused exponential height formula into a three layer system including sun influence, linear colors, adjusted sky coloring and alternative blending/balancing
- Updated Sky defaults, new fog settings and adjustments to Tonemapping exposure for point light shadows.
- Added a PreTonemapper for no-dxvk AMD users who can’t use full tonemapping, needs ideally a weather mod with minimal cinematic modifiers and no tints.

TheFireKahuna committed:
Linearise more shaders, updated Lottes, reworked volumetric fog

- Updated Lottes with vtastek's new changes
- Optimised Lottes and the tonemapping shaders in an unsuccessful attempt to help AMD users
- Linearised Exterior Shadows, Ambient Occlusion, Underwater, Wet World
- Fixed Godrays, had too much linearisation
- Moved AO to after Shadows as linear space blending is different (needs testing)

TheFireKahuna committed:
Reworked Volumetric Fog

- Reworked Volumetric Fog with the unused exponential height formula, sun influence, double layering, linear colors, adjusted sky coloring and alternative blending/balancing
- Defaults reset to 1.0 for Fog and Skies

TheFireKahuna committed:
Finish new linear fog

Linearise all relevant shaders, test interior shaders, rain and snow before release

TheFireKahuna committed:
Update NewVegasReloaded.dll.defaults.toml

TheFireKahuna committed:
Final fog adjustments

- Cleanup some legacy commented out code
- Fix the sun factor causing weird coloring
- Add separate configurable height falloffs for all three forms of fog
- Rework how distance falloffs occur, much more configurable with more desired behaviour
- Rebalanced all values again for a better mix
- Reduced night influence on simple fog blend

TheFireKahuna committed:
Adjustments for linearised NVR

- Linearise the bloom shader, vanilla bloom should be replaced
- Balance snow accumulation for linear color

TheFireKahuna commented:
Should be all ready for a merge

TheFireKahuna committed:
Extra fixes

- Slightly lower base Lottes contrast (same as last nightly)
- Use SimpleFogDist to instead have an Exterior/Interior value to neutralise fog color with
- Adjust default NVR settings, specular strength adjustment for most weather mods, more exposure for Point Lights and adjusted shadow darkness. Also set Bloom Exponent at default

TheFireKahuna committed:
Reduce clouds brightness boost

TheFireKahuna committed:
Add an alternative PreTonemapper for no-DXVK AMD

Applies before the actual tonemapper, meaning Cinematic modifiers would be post tonemap and tints applied atop. At the least it corrects overblown skies and lighting, allowing NVR to be semi-playable without excessive tuning

mchaptel commented on Extra/Shaders/NewVegasReloaded/Shaders/ISHDRBLENDINSHADERCIN.pso.hlsl:
```diff
@@ -49,36 +49,32 @@ struct VS_OUTPUT {
     
-    final = lerp(final.rgb, Cinematic.z * (final.rgb - Cinematic.y) + Cinematic.y, TESR_HDRData.z * TESR_ToneMapping.y); // apply contrast from Cinematic, scaled by modifier
-    final = lerp(final.rgb, Fade.rgb, Fade.a); // apply night eye and fade, gamma-space but vanilla accurate
+    final.rgb = lerp(final.rgb, Cinematic.w * (final.rgb - Cinematic.y) + Cinematic.y, cinematicScalar * TESR_ToneMapping.y); // apply contrast from Cinematic, scaled by modifier
```
are you sure this is cinematic.w here?

mchaptel replied:
https://github.com/mchaptel/TESReloaded10/commit/71b7c1108f3e369f01da33818eaf23b8003598a7#diff-ae7d9f078a232fd5a48ee69ba0b5e62e6146b76ba65caced9b07045f875e8641R79

looks from the base shader that w is the brightness maybe? Like a base exposure multiplier before contrast is applied

TheFireKahuna replied:
Yeah mistake there, should be Cinematic.z. Fixed

mchaptel commented on Extra/Shaders/NewVegasReloaded/Shaders/ISHDRBLENDINSHADERCINAM.pso.hlsl:
```diff
@@ -51,37 +51,33 @@ struct VS_OUTPUT {
     
-    final = lerp(final.rgb, Cinematic.z * (final.rgb - Cinematic.y) + Cinematic.y, TESR_HDRData.z * TESR_ToneMapping.y); // apply contrast from Cinematic, scaled by modifier
-    final = lerp(final.rgb, Fade.rgb, Fade.a); // apply night eye and fade, gamma-space but vanilla accurate
+    final.rgb = lerp(final.rgb, Cinematic.w * (final.rgb - Cinematic.y) + Cinematic.y, cinematicScalar * TESR_ToneMapping.y); // apply contrast from Cinematic, scaled by modifier
```
same question here as https://github.com/mchaptel/TESReloaded10/pull/33/files#diff-ae7d9f078a232fd5a48ee69ba0b5e62e6146b76ba65caced9b07045f875e8641R74

TheFireKahuna replied:
Same mistake, also fixed. .z was also being used as exposure, which is now .w

mchaptel commented on Extra/Shaders/NewVegasReloaded/Shaders/Includes/Tonemapping.hlsl:
```diff
@@ -215,47 +215,91 @@ float ColToneC(float hdrMax, float contrast, float shoulder, float midIn, float
 }
 #define CMAX 1.6e+6f
+#define EPS 1e-6f
 float3 VTLottes(float3 color, float contrast, float midOut, float midIn, float hdrMax, float shoulder, float crossTalk)
```
are we no longer using VTLottes? It could be good here to have a link to vtastek's gist for sourcing

TheFireKahuna replied:
Renamed AMDLottes back to VTLottes and removed the larger non-optimised code I commented out, since AMDLottes is just optimised VTLottes. Should restore the credits and trim the code

mchaptel commented on Extra/Shaders/NewVegasReloaded/Shaders/SKYTEX.pso.hlsl:
```diff
@@ -156,12 +156,12 @@ VS_OUTPUT main(VS_INPUT IN) {
         cloudTint = lerp(cloudTint, white.rgb, sunHeight * isDayTime); // tint the clouds less when the sun is high in the sky
 
-        finalColor.rgb *= lerp(1.0, cloudTint * TESR_CloudData.w * 1.5, isDayTime); // cancel tint at night
+        finalColor.rgb *= lerp(1.0, cloudTint * TESR_CloudData.w * 1.333 * TESR_DebugVar.z, isDayTime); // cancel tint at night
```
it would be good to remove the debug vars from the shaders before release

TheFireKahuna replied:
True, removed

mchaptel commented on Extra/Shaders/NewVegasReloaded/Shaders/SKYTEX.pso.hlsl:
```diff
@@ -122,15 +122,15 @@ VS_OUTPUT main(VS_INPUT IN) {
 
     // shade clouds 
-    float greyScale = lerp(luma(finalColor), 1.0, saturate(TESR_CloudData.w));
+    float greyScale = lerp(luma(finalColor), 1.0, saturate(TESR_CloudData.w * TESR_DebugVar.w));
```
it would be good to remove the debug vars from the shaders before release

TheFireKahuna replied:
Also removed

mchaptel reviewed (changes requested):
Looks great! With the length of the diff, I placed some questions/remarks directly as review comments for ease. Let me know what you thinl

TheFireKahuna committed:
Changes as per code review

- Cleanup optimised VTLottes
- Removed unused debug vars
- Fixed the cinematic curve modifiers being swapped
- Add missing ImageAdjust fixes
- Minor optimisation to tonemapper
- Prevent PreTonemapper from breaking on changing Tonemap Mode (locked de-facto to VTLottes)

TheFireKahuna committed:
Fix Wetworld and linearise Snow

- Fixed delinearisation not occuring at points in the Wetworld shader
- Linearise Snow shader

TheFireKahuna committed:
Preserve prior method of computing point shadow darkness

To try and avoid any issues
